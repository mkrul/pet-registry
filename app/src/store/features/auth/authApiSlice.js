import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { setUser, clearUser, setLoading, setError, updateLastActivity } from "./authSlice";
import { addNotification } from "../notifications/notificationsSlice";

export const authApiSlice = createApi({
  reducerPath: "authApi",
  baseQuery: fetchBaseQuery({
    baseUrl: "/api",
    credentials: "include",
    prepareHeaders: headers => {
      const token = document.querySelector('meta[name="csrf-token"]')?.getAttribute("content");
      if (token) {
        headers.set("X-CSRF-Token", token);
      }
      headers.set("Accept", "application/json");
      headers.set("Content-Type", "application/json");
      return headers;
    }
  }),
  tagTypes: ["Auth"],
  endpoints: builder => ({
    login: builder.mutation({
      query: credentials => ({
        url: "/login",
        method: "POST",
        body: credentials,
        credentials: "include"
      }),
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        dispatch(setLoading(true));
        try {
          const { data } = await queryFulfilled;
          dispatch(setUser(data.user));
          dispatch(
            addNotification({
              type: "SUCCESS",
              message: data.message || "Logged in successfully"
            })
          );
        } catch (err) {
          const errorMessage = err?.error?.data?.error || err?.data?.error || "Login failed";
          dispatch(setError(errorMessage));
          dispatch(clearUser());
          dispatch(
            addNotification({
              type: "ERROR",
              message: errorMessage
            })
          );
        } finally {
          dispatch(setLoading(false));
        }
      }
    }),
    logout: builder.mutation({
      query: () => ({
        url: "/logout",
        method: "DELETE",
        credentials: "include"
      }),
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        try {
          await queryFulfilled;
          localStorage.clear();
          sessionStorage.clear();
          dispatch(clearUser());
          dispatch(authApiSlice.util.resetApiState());
          window.location.href = "/login";
        } catch (err) {
          dispatch(setError("Logout failed"));
        }
      }
    }),
    getCurrentUser: builder.query({
      query: () => ({
        url: "/current_user",
        credentials: "include"
      }),
      async onQueryStarted(_, { dispatch, queryFulfilled }) {
        dispatch(setLoading(true));
        try {
          const { data } = await queryFulfilled;
          if (data?.user) {
            dispatch(setUser(data.user));
          }
        } catch (err) {
          if (err?.status === 401) {
            dispatch(clearUser());
          }
        } finally {
          dispatch(setLoading(false));
        }
      }
    }),
    signUp: builder.mutation({
      query: credentials => ({
        url: "/sign_up",
        method: "POST",
        body: credentials
      }),
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        dispatch(setLoading(true));
        try {
          const { data } = await queryFulfilled;
          dispatch(setUser(data.user));
        } catch (err) {
          dispatch(setError(err?.data?.message || "Sign up failed"));
          dispatch(clearUser());
        } finally {
          dispatch(setLoading(false));
        }
      }
    }),
    pollSession: builder.mutation({
      query: () => ({
        url: "/api/sessions/poll",
        method: "POST",
        credentials: "include"
      }),
      async onQueryStarted(_, { dispatch }) {
        dispatch(updateLastActivity());
      }
    }),
    updateUser: builder.mutation({
      query: (userData) => ({
        url: "/profile",
        method: "PATCH",
        body: { user: userData },
        credentials: "include"
      }),
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        try {
          const { data } = await queryFulfilled;
          if (data?.user) {
            dispatch(setUser(data.user));
          }
          dispatch(
            addNotification({
              type: "SUCCESS",
              message: data.message || "Profile updated successfully"
            })
          );
        } catch (err) {
          const errorMessage = err?.error?.data?.error || err?.error?.data?.message || "Update failed";
          dispatch(
            addNotification({
              type: "ERROR",
              message: errorMessage
            })
          );
          throw err;
        }
      }
    }),
    changePassword: builder.mutation({
      query: (passwordData) => ({
        url: "/change_password",
        method: "PATCH",
        body: { user: passwordData },
        credentials: "include"
      }),
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        try {
          const { data } = await queryFulfilled;
          if (data?.user) {
            dispatch(setUser(data.user));
          }
          dispatch(
            addNotification({
              type: "SUCCESS",
              message: data.message || "Password changed successfully"
            })
          );
        } catch (err) {
          const errorMessage = err?.error?.data?.error || err?.error?.data?.message || "Password change failed";
          dispatch(
            addNotification({
              type: "ERROR",
              message: errorMessage
            })
          );
          throw err;
        }
      }
    })
  })
});

export const {
  useLoginMutation,
  useLogoutMutation,
  useGetCurrentUserQuery,
  useSignUpMutation,
  usePollSessionMutation,
  useUpdateUserMutation,
  useChangePasswordMutation
} = authApiSlice;
