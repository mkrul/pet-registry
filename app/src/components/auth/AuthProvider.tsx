import React, { useEffect } from "react";
import { useGetCurrentUserQuery } from "../../redux/features/auth/authApiSlice";
import { useAppDispatch } from "../../redux/hooks";
import {
  setUser,
  clearUser,
  setAuthLoading,
  checkAuthTimeout
} from "../../redux/features/auth/authSlice";

interface AuthProviderProps {
  children: React.ReactNode;
}

const AuthProvider: React.FC<AuthProviderProps> = ({ children }) => {
  const dispatch = useAppDispatch();

  // Fetch current user with polling enabled
  const { data, isLoading } = useGetCurrentUserQuery(undefined, {
    // Poll every 5 minutes to keep session alive
    pollingInterval: 5 * 60 * 1000
  });

  // Initial auth check and periodic updates
  useEffect(() => {
    if (!isLoading) {
      if (data?.user) {
        dispatch(setUser(data.user));
      } else {
        dispatch(clearUser());
      }
    } else {
      dispatch(setAuthLoading());
    }
  }, [data, isLoading, dispatch]);

  // Check auth timeout periodically
  useEffect(() => {
    const checkTimeout = () => {
      dispatch(checkAuthTimeout());
    };

    // Check every minute
    const interval = setInterval(checkTimeout, 60 * 1000);

    // Check when tab becomes visible
    const handleVisibilityChange = () => {
      if (document.visibilityState === "visible") {
        checkTimeout();
      }
    };

    document.addEventListener("visibilitychange", handleVisibilityChange);

    // Initial check
    checkTimeout();

    return () => {
      clearInterval(interval);
      document.removeEventListener("visibilitychange", handleVisibilityChange);
    };
  }, [dispatch]);

  // Render children without additional wrapper to avoid unnecessary DOM nesting
  return <>{children}</>;
};

export default AuthProvider;
