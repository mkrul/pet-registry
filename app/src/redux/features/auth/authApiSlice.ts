import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { setUser, clearUser } from "./authSlice";
import { AuthResponse, SignUpRequest } from "../../../types/auth/AuthApiSlice";

export const authApiSlice = createApi({
  reducerPath: "authApi",
  baseQuery: fetchBaseQuery({
    baseUrl: "/api",
    credentials: "include",
    prepareHeaders: headers => {
      headers.set("Accept", "application/json");
      headers.set("Content-Type", "application/json");
      return headers;
    }
  }),
  tagTypes: ["Auth"],
  keepUnusedDataFor: 3600,
  endpoints: builder => ({
    login: builder.mutation<AuthResponse, { user: { email: string; password: string } }>({
      query: credentials => ({
        url: "/login",
        method: "POST",
        body: credentials
      }),
      invalidatesTags: ["Auth"],
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        try {
          const { data } = await queryFulfilled;
          dispatch(setUser(data.user));
        } catch (err) {
          dispatch(clearUser());
        }
      },
      transformErrorResponse: (response: { status: number; data: any }) => ({
        status: response.status,
        data: {
          message: response.data?.message
        }
      })
    }),
    logout: builder.mutation<{ message: string }, void>({
      query: () => ({
        url: "/logout",
        method: "DELETE"
      }),
      invalidatesTags: ["Auth"],
      async onQueryStarted(_, { dispatch, queryFulfilled }) {
        try {
          await queryFulfilled;
          dispatch(clearUser());
        } catch (err) {
          dispatch(clearUser());
        }
      },
      transformErrorResponse: (response: { status: number; data: any }) => ({
        status: response.status,
        data: {
          message: response.data?.message
        }
      })
    }),
    getCurrentUser: builder.query<AuthResponse, void>({
      query: () => ({
        url: "/current_user",
        method: "GET"
      }),
      keepUnusedDataFor: 3600,
      async onQueryStarted(_, { dispatch, queryFulfilled }) {
        try {
          const { data } = await queryFulfilled;
          if (data?.user) {
            dispatch(setUser(data.user));
          } else {
            dispatch(clearUser());
          }
        } catch (err) {
          dispatch(clearUser());
        }
      },
      providesTags: ["Auth"],
      transformErrorResponse: (response: { status: number; data: any }) => ({
        status: response.status,
        data: {
          message: response.data?.message
        }
      })
    }),
    signUp: builder.mutation<AuthResponse, SignUpRequest>({
      query: credentials => ({
        url: "/auth/registration",
        method: "POST",
        body: credentials
      }),
      invalidatesTags: ["Auth"],
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        try {
          const { data } = await queryFulfilled;
          dispatch(setUser(data.user));
        } catch (err) {
          dispatch(clearUser());
        }
      },
      transformErrorResponse: (response: { status: number; data: any }) => ({
        status: response.status,
        data: {
          message: response.data?.message
        }
      })
    })
  })
});

export const { useLoginMutation, useLogoutMutation, useGetCurrentUserQuery, useSignUpMutation } =
  authApiSlice;
