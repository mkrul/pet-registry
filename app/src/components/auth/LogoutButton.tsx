import React from "react";
import { useLogoutMutation } from "../../redux/features/auth/authApiSlice";
import { useAppDispatch } from "../../redux/hooks";
import { clearUser } from "../../redux/features/auth/authSlice";

const LogoutButton: React.FC = () => {
  const dispatch = useAppDispatch();
  const [logout] = useLogoutMutation();

  const handleLogout = async () => {
    try {
      const response = await logout().unwrap();
      console.log(response.message);
      dispatch(clearUser());
      // Redirect to login if needed
    } catch (err) {
      console.error("Logout failed:", err);
    }
  };

  return <button onClick={handleLogout}>Logout</button>;
};

export default LogoutButton;
