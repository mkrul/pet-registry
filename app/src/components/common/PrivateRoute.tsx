import React, { useEffect } from "react";
import { Navigate, Outlet, useLocation } from "react-router-dom";
import { useAppSelector } from "../../redux/hooks";
import { useGetCurrentUserQuery } from "../../redux/features/auth/authApiSlice";
import Spinner from "../shared/Spinner";

const PrivateRoute: React.FC = () => {
  const user = useAppSelector(state => state.auth.user);
  const location = useLocation();

  const { isLoading, error } = useGetCurrentUserQuery(undefined, {
    skip: !!user,
    refetchOnMountOrArgChange: 3600
  });

  if (isLoading) {
    return <Spinner />;
  }

  if (!user) {
    return <Navigate to="/login" replace state={{ from: location }} />;
  }

  return <Outlet />;
};

export default PrivateRoute;
