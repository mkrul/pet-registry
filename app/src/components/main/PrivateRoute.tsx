import React from "react";
import { Navigate, Outlet, useLocation } from "react-router-dom";
import { useAppSelector } from "../../redux/hooks";
import { useGetCurrentUserQuery } from "../../redux/features/auth/authApiSlice";
import Spinner from "../common/Spinner";

const PrivateRoute: React.FC = () => {
  const user = useAppSelector(state => state.auth.user);
  const location = useLocation();

  const { isLoading, error } = useGetCurrentUserQuery(undefined, {
    skip: !!user,
    refetchOnMountOrArgChange: 3600
  });

  if (isLoading) {
    return <Spinner />;
  }

  if (!user || error) {
    return <Navigate to="/login" replace state={{ from: location }} />;
  }

  return <Outlet />;
};

export default PrivateRoute;
