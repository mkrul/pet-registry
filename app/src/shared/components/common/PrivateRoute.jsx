import React from "react";
import { Navigate, Outlet, useLocation } from "react-router-dom";
import { useAppSelector } from "../../../store/hooks.js";
import { useGetCurrentUserQuery } from "../../../store/features/auth/authApiSlice.js";
import Spinner from "./Spinner";

const PrivateRoute = () => {
  const user = useAppSelector(state => state.auth.user);
  const location = useLocation();

  const { isLoading, error } = useGetCurrentUserQuery(undefined, {
    skip: !!user,
    refetchOnMountOrArgChange: 3600
  });


  if (isLoading && !user) {
    return <Spinner />;
  }

  if (!user || error) {
    return <Navigate to="/login" replace state={{ from: location }} />;
  }

  return <Outlet />;
};

export default PrivateRoute;
