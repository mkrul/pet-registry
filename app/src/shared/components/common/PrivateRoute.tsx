import React from "react";
import { Navigate, Outlet, useLocation } from "react-router-dom";
import { useAppSelector } from "../../../store/hooks";
import { useGetCurrentUserQuery } from "../../../store/features/auth/authApiSlice";
import Spinner from "./Spinner";

const PrivateRoute: React.FC = () => {
  const user = useAppSelector(state => state.auth.user);
  const location = useLocation();

  const { isLoading, error } = useGetCurrentUserQuery(undefined, {
    skip: !!user,
    refetchOnMountOrArgChange: 3600
  });


  if (isLoading && !user) {
    return <Spinner />;
  }

  if (!user || error) {
    return <Navigate to="/login" replace state={{ from: location }} />;
  }

  return <Outlet />;
};

export default PrivateRoute;
