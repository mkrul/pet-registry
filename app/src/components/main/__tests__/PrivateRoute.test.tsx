import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/vitest";
import { MemoryRouter, Route, Routes, useLocation } from "react-router-dom";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import PrivateRoute from "../PrivateRoute";
import authReducer from "../../../redux/features/auth/authSlice";
import { User } from "../../../types/auth/User";
import { useGetCurrentUserQuery } from "../../../redux/features/auth/authApiSlice";

vi.mock("../../../redux/features/auth/authApiSlice", () => ({
  useGetCurrentUserQuery: vi.fn()
}));

const ProtectedComponent = () => <div data-testid="protected">Protected Content</div>;
const LoginPage = () => {
  const location = useLocation();
  return (
    <div data-testid="login">
      Login Page
      <span data-testid="location-state">{JSON.stringify(location.state)}</span>
    </div>
  );
};

vi.mock("../../common/Spinner", () => ({
  default: () => <div data-testid="spinner">Loading...</div>
}));

const renderPrivateRoute = (
  initialState = { auth: { user: null as User | null } },
  initialRoute = "/protected"
) => {
  const store = configureStore({
    reducer: { auth: authReducer },
    preloadedState: initialState
  });

  return {
    ...render(
      <Provider store={store}>
        <MemoryRouter initialEntries={[initialRoute]}>
          <Routes>
            <Route element={<PrivateRoute />}>
              <Route path="/protected" element={<ProtectedComponent />} />
            </Route>
            <Route path="/login" element={<LoginPage />} />
          </Routes>
        </MemoryRouter>
      </Provider>
    ),
    store
  };
};

describe("PrivateRoute", () => {
  beforeEach(() => {
    vi.mocked(useGetCurrentUserQuery).mockReturnValue({
      data: null,
      isLoading: false,
      error: null,
      refetch: vi.fn()
    } as any);
  });

  describe("Authentication States", () => {
    it("shows loading spinner while checking auth status", () => {
      vi.mocked(useGetCurrentUserQuery).mockReturnValue({
        data: null,
        isLoading: true,
        error: null,
        refetch: vi.fn()
      } as any);

      const { container } = renderPrivateRoute();
      expect(screen.getByTestId("spinner")).toBeDefined();
      expect(container).toMatchSnapshot();
    });

    it("redirects to login when user is not authenticated", async () => {
      vi.mocked(useGetCurrentUserQuery).mockReturnValue({
        data: null,
        isLoading: false,
        error: null,
        refetch: vi.fn()
      } as any);

      renderPrivateRoute();

      await waitFor(() => {
        expect(screen.getByTestId("login")).toBeDefined();
      });
      expect(screen.getByTestId("location-state")).toHaveTextContent("/protected");
    });

    it("allows access to protected route when user is authenticated", async () => {
      const mockUser: User = { id: 1, email: "test@example.com" };

      vi.mocked(useGetCurrentUserQuery).mockReturnValue({
        data: mockUser,
        isLoading: false,
        error: null,
        refetch: vi.fn()
      } as any);

      renderPrivateRoute({ auth: { user: mockUser } });

      await waitFor(() => {
        expect(screen.getByTestId("protected")).toBeDefined();
      });
    });
  });

  describe("Error Handling", () => {
    it("redirects to login on authentication error", async () => {
      vi.mocked(useGetCurrentUserQuery).mockReturnValue({
        data: null,
        isLoading: false,
        error: { status: 401, data: { message: "Unauthorized" } },
        refetch: vi.fn()
      } as any);

      renderPrivateRoute();

      await waitFor(() => {
        expect(screen.getByTestId("login")).toBeDefined();
        expect(screen.getByTestId("location-state")).toHaveTextContent("/protected");
      });
    });
  });

  describe("Route State Preservation", () => {
    it("preserves attempted route in location state when redirecting", async () => {
      const testRoute = "/protected?param=test";

      vi.mocked(useGetCurrentUserQuery).mockReturnValue({
        data: null,
        isLoading: false,
        error: null,
        refetch: vi.fn()
      } as any);

      renderPrivateRoute(undefined, testRoute);

      await waitFor(() => {
        expect(screen.getByTestId("login")).toBeDefined();
        const locationState = JSON.parse(screen.getByTestId("location-state").textContent || "{}");
        expect(locationState.from.pathname).toBe("/protected");
        expect(locationState.from.search).toBe("?param=test");
      });
    });
  });
});
