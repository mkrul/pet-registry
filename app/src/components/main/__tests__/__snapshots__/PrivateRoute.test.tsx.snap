// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PrivateRoute > Authentication States > shows loading spinner while checking auth status 1`] = `
<div>
  <div
    data-testid="spinner"
  >
    Loading...
  </div>
</div>
`;
